From d26f61a6ebbee18755164265e3c5fc3d0feb0729 Mon Sep 17 00:00:00 2001
From: Kiet Pham <kiet.pham.xb@renesas.com>
Date: Tue, 13 Aug 2024 14:53:37 +0700
Subject: [PATCH 4/4] omxvideodec: Override sink_query

gst-omx decoder does not work as same as video filter / converter. The
difference of resolution between sink pad and source pad is not allowed.

This patch is a work-around to help gst-omx decoder pass the negotiation phase
if it uses scaling feature. It will return the template caps of sink pad to
upstream instead of a proxy caps.

Signed-off-by: Kiet Pham <kiet.pham.xb@renesas.com>
---
 omx/gstomxvideodec.c | 38 ++++++++++++++++++++++++++++++++++++++
 1 file changed, 38 insertions(+)

diff --git a/omx/gstomxvideodec.c b/omx/gstomxvideodec.c
index d56c74d..82d31ca 100644
--- a/omx/gstomxvideodec.c
+++ b/omx/gstomxvideodec.c
@@ -82,6 +82,8 @@ static gboolean gst_omx_video_dec_decide_allocation (GstVideoDecoder * bdec,
     GstQuery * query);
 static gboolean gst_omx_video_dec_propose_allocation (GstVideoDecoder * bdec,
     GstQuery * query);
+static gboolean gst_omx_video_dec_sink_query (GstVideoDecoder * decoder,
+    GstQuery * query);

 static GstFlowReturn gst_omx_video_dec_drain (GstVideoDecoder * decoder);

@@ -328,6 +330,8 @@ gst_omx_video_dec_class_init (GstOMXVideoDecClass * klass)
       GST_DEBUG_FUNCPTR (gst_omx_video_dec_decide_allocation);
   video_decoder_class->propose_allocation =
       GST_DEBUG_FUNCPTR (gst_omx_video_dec_propose_allocation);
+  video_decoder_class->sink_query =
+      GST_DEBUG_FUNCPTR (gst_omx_video_dec_sink_query);

   klass->cdata.type = GST_OMX_COMPONENT_TYPE_FILTER;
   klass->cdata.default_src_template_caps =
@@ -3959,6 +3963,40 @@ gst_omx_video_dec_propose_allocation (GstVideoDecoder * bdec, GstQuery * query)
       (gst_omx_video_dec_parent_class)->propose_allocation (bdec, query);
 }

+static gboolean
+gst_omx_video_dec_sink_query (GstVideoDecoder * decoder, GstQuery * query)
+{
+  GstOMXVideoDec *self = GST_OMX_VIDEO_DEC (decoder);
+  GstPad *pad = GST_VIDEO_DECODER_SINK_PAD (self);
+
+  GST_LOG_OBJECT (self, "handling query: %" GST_PTR_FORMAT, query);
+
+  if (self->dec) {
+    if (self->enable_scale == TRUE && GST_QUERY_TYPE (query) == GST_QUERY_CAPS) {
+      GstCaps *filter, *caps;
+
+      gst_query_parse_caps (query, &filter);
+      caps = gst_video_decoder_proxy_getcaps (decoder, NULL, filter);
+
+      if (filter == NULL || gst_caps_is_empty (caps)) {
+        GST_DEBUG_OBJECT(self,
+                         "No available proxy caps, using sink pad instead");
+        gst_caps_replace (&caps, gst_pad_get_pad_template_caps (pad));
+      }
+
+      GST_LOG_OBJECT (self, "Returning caps %" GST_PTR_FORMAT, caps);
+      gst_query_set_caps_result (query, caps);
+      gst_caps_unref (caps);
+
+      return TRUE;
+    }
+  }
+
+  return
+    GST_VIDEO_DECODER_CLASS
+    (gst_omx_video_dec_parent_class)->sink_query (decoder, query);
+}
+
 static gint
 gst_omx_video_dec_calculate_scale_ratio (gint new_length, gint length)
 {
--
2.25.1

